substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/AbstractDynamicClassGenerationSupport.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.agent;
+
+import com.oracle.svm.core.util.JavaClassUtil;
+import static com.oracle.svm.jvmtiagentbase.Support.jniFunctions;
+import static com.oracle.svm.jvmtiagentbase.Support.getMethodFullNameAtFrame;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import com.oracle.svm.jni.nativeapi.JNIMethodId;
+import org.graalvm.nativeimage.c.type.CCharPointer;
+import org.graalvm.nativeimage.c.type.CTypeConversion;
+import org.graalvm.nativeimage.c.type.VoidPointer;
+import org.graalvm.word.WordFactory;
+
+import com.oracle.svm.jni.nativeapi.JNIEnvironment;
+import com.oracle.svm.jni.nativeapi.JNIObjectHandle;
+
+public abstract class AbstractDynamicClassGenerationSupport {
+
+    protected JNIEnvironment jni;
+    protected JNIObjectHandle callerClass;
+    protected final String generatedClassName;
+    protected TraceWriter traceWriter;
+    protected NativeImageAgent agent;
+    protected int generatedClassHashCode = 0;
+    protected byte[] values = null;
+
+    private static Path dynclassDumpDir = null;
+    private static final Path DEFAULT_DUMP = Paths.get("dynClass");
+    protected String callerMethod;
+
+    public static void initDynClassDumpDir(String dir) {
+        dynclassDumpDir = Paths.get(dir);
+    }
+
+    public static AbstractDynamicClassGenerationSupport getClassLoaderDefineClassSupport(JNIEnvironment jni, JNIObjectHandle callerClass,
+                    String generatedClassName, TraceWriter traceWriter, NativeImageAgent agent) throws IOException {
+        return new ClassLoaderDefineClassSupport(jni, callerClass, generatedClassName, traceWriter, agent);
+    }
+
+    protected AbstractDynamicClassGenerationSupport(JNIEnvironment jni, JNIObjectHandle callerClass,
+                    String generatedClassName, TraceWriter traceWriter, NativeImageAgent agent) throws IOException {
+        if (dynclassDumpDir == null) {
+            System.out.println("Warning: dynamic-class-dump-dir= was not set in -agentlib:native-image-agent=. Dynamically generated classes will be dumped to the default location:" +
```

Comment:
I would avoid printing in the agent as this could alter the behavior of the program. In other places the agent uses `System.err`.

Reply:
fixed